File: integration_tests/java_client_tests_real/client-app/src/test/java/dev/graphlink/test/ErrorCaptureTest.java
package dev.graphlink.test;

import dev.graphlink.test.generated.client.GraphLinkClient;
import dev.graphlink.test.generated.types.CreateUserOrErrorsFullResponse;
import dev.graphlink.test.generated.types.FindUserOrErrorsFullResponse;
import dev.graphlink.test.generated.types.GetUserOrErrorsFullResponse;
import dev.graphlink.test.generated.inputs.AddressInput;
import dev.graphlink.test.generated.inputs.CreateUserInput;
import dev.graphlink.test.generated.enums.UserStatus;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;

import static dev.graphlink.test.Fixtures.*;
import static org.junit.jupiter.api.Assertions.*;

class ErrorCaptureTest {

    private GraphLinkClient client;

    private static final AddressInput kAddress = AddressInput.builder()
            .street("1 Main St").city("Springfield").country("US").zip("12345").build();
    private static final CreateUserInput kCreateInput = CreateUserInput.builder()
            .name("Alice").email("alice@test.com").status(UserStatus.ACTIVE).address(kAddress).build();
    private static final CreateUserInput kErrorInput = CreateUserInput.builder()
            .name("error-user").email("e@test.com").status(UserStatus.ACTIVE).address(kAddress).build();

    @BeforeEach
    void setUp() {
        client = newRealClient();
    }

    // ── getUserOrErrors — success ──────────────────────────────────────────────

    @Nested
    class GetUserOrErrorsSuccess {

        @Test
        void returnsFullResponse() {
            assertNotNull(client.queries.getUserOrErrors("user-1"));
        }

        @Test
        void data_isNotNull() {
            assertNotNull(client.queries.getUserOrErrors("user-1").getData());
        }

        @Test
        void data_user_hasCorrectId() {
            assertEquals("user-1",
                    client.queries.getUserOrErrors("user-1").getData().getGetUserOrErrors().getId());
        }

        @Test
        void errors_isNullOrEmpty_onSuccess() {
            GetUserOrErrorsFullResponse res = client.queries.getUserOrErrors("user-1");
            assertTrue(res.getErrors() == null || res.getErrors().isEmpty());
        }

        @Test
        void doesNotThrow_onSuccess() {
            assertDoesNotThrow(() -> client.queries.getUserOrErrors("user-1"));
        }
    }

    // ── getUserOrErrors — server-side GraphQL error (unknown id) ──────────────

    @Nested
    class GetUserOrErrorsWithErrors {

        @Test
        void doesNotThrow_whenServerReturnsError() {
            assertDoesNotThrow(() -> client.queries.getUserOrErrors("unknown-id"));
        }

        @Test
        void errors_isNonEmpty_forUnknownId() {
            GetUserOrErrorsFullResponse res = client.queries.getUserOrErrors("unknown-id");
            assertNotNull(res.getErrors());
            assertFalse(res.getErrors().isEmpty());
        }

        @Test
        void errors_hasMessage_forUnknownId() {
            GetUserOrErrorsFullResponse res = client.queries.getUserOrErrors("unknown-id");
            assertNotNull(res.getErrors().get(0).getMessage());
            assertFalse(res.getErrors().get(0).getMessage().isBlank());
        }
    }

    // ── findUserOrErrors — null return (non-error) ────────────────────────────

    @Nested
    class FindUserOrErrorsNullReturn {

        @Test
        void doesNotThrow_forUnknownId() {
            assertDoesNotThrow(() -> client.queries.findUserOrErrors("user-99"));
        }

        @Test
        void data_fieldIsNull_forUnknownId() {
            FindUserOrErrorsFullResponse res = client.queries.findUserOrErrors("user-99");
            assertTrue(res.getData() == null
                    || res.getData().getFindUserOrErrors() == null);
        }

        @Test
        void errors_isNullOrEmpty_forUnknownId() {
            FindUserOrErrorsFullResponse res = client.queries.findUserOrErrors("user-99");
            assertTrue(res.getErrors() == null || res.getErrors().isEmpty());
        }
    }

    // ── findUserOrErrors — server error ───────────────────────────────────────

    @Nested
    class FindUserOrErrorsWithErrors {

        @Test
        void doesNotThrow_forErrorId() {
            assertDoesNotThrow(() -> client.queries.findUserOrErrors("error-id"));
        }

        @Test
        void errors_isNonEmpty_forErrorId() {
            FindUserOrErrorsFullResponse res = client.queries.findUserOrErrors("error-id");
            assertNotNull(res.getErrors());
            assertFalse(res.getErrors().isEmpty());
        }
    }

    // ── findUserOrErrors — success ────────────────────────────────────────────

    @Nested
    class FindUserOrErrorsSuccess {

        @Test
        void data_isPresent_forKnownId() {
            FindUserOrErrorsFullResponse res = client.queries.findUserOrErrors("user-1");
            assertNotNull(res.getData());
            assertNotNull(res.getData().getFindUserOrErrors());
            assertEquals("user-1", res.getData().getFindUserOrErrors().getId());
        }
    }

    // ── Non-captureErrors query throws on server error ────────────────────────

    @Nested
    class NonCaptureErrorsThrows {

        @Test
        void getUser_throws_forUnknownId() {
            assertThrows(RuntimeException.class, () -> client.queries.getUser("unknown-id"));
        }
    }

    // ── createUserOrErrors mutation ───────────────────────────────────────────

    @Nested
    class CreateUserOrErrorsMutation {

        @Test
        void success_dataIsPresent() {
            CreateUserOrErrorsFullResponse res = client.mutations.createUserOrErrors(kCreateInput);
            assertNotNull(res.getData());
            assertNotNull(res.getData().getCreateUserOrErrors());
        }

        @Test
        void success_noErrors() {
            CreateUserOrErrorsFullResponse res = client.mutations.createUserOrErrors(kCreateInput);
            assertTrue(res.getErrors() == null || res.getErrors().isEmpty());
        }

        @Test
        void doesNotThrow_whenServerReturnsError() {
            assertDoesNotThrow(() -> client.mutations.createUserOrErrors(kErrorInput));
        }

        @Test
        void errors_arePopulated_whenServerReturnsError() {
            CreateUserOrErrorsFullResponse res = client.mutations.createUserOrErrors(kErrorInput);
            assertNotNull(res.getErrors());
            assertFalse(res.getErrors().isEmpty());
        }
    }
}

File: integration_tests/kotlin_client_tests_real/client-app/src/test/kotlin/dev/graphlink/test/ErrorCaptureTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.enums.UserStatus
import dev.graphlink.test.generated.inputs.AddressInput
import dev.graphlink.test.generated.inputs.CreateUserInput
import dev.graphlink.test.generated.types.CreateUserOrErrorsFullResponse
import dev.graphlink.test.generated.types.FindUserOrErrorsFullResponse
import dev.graphlink.test.generated.types.GetUserOrErrorsFullResponse
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.newRealClient

class ErrorCaptureTest {

    private lateinit var client: GraphLinkClient

    private val kAddress = AddressInput("1 Main St", "Springfield", "US", "12345")
    private val kCreateInput = CreateUserInput(
        name = "Alice", email = "alice@test.com",
        status = UserStatus.ACTIVE, address = kAddress,
    )
    private val kErrorInput = CreateUserInput(
        name = "error-user", email = "e@test.com",
        status = UserStatus.ACTIVE, address = kAddress,
    )

    @BeforeTest
    fun setUp() {
        client = newRealClient()
    }

    // ── getUserOrErrors — success ──────────────────────────────────────────────

    @Nested
    inner class GetUserOrErrorsSuccess {

        @Test
        fun returnsFullResponse() = runTest {
            assertNotNull(client.queries.getUserOrErrors("user-1"))
        }

        @Test
        fun data_isNotNull() = runTest {
            assertNotNull(client.queries.getUserOrErrors("user-1").data)
        }

        @Test
        fun data_user_hasCorrectId() = runTest {
            assertEquals("user-1",
                client.queries.getUserOrErrors("user-1").data!!.getUserOrErrors.id)
        }

        @Test
        fun errors_isNullOrEmpty_onSuccess() = runTest {
            val res: GetUserOrErrorsFullResponse = client.queries.getUserOrErrors("user-1")
            assertTrue(res.errors == null || res.errors!!.isEmpty())
        }

        @Test
        fun doesNotThrow_onSuccess() = runTest {
            client.queries.getUserOrErrors("user-1")
        }
    }

    // ── getUserOrErrors — server-side error (unknown id) ──────────────────────

    @Nested
    inner class GetUserOrErrorsWithErrors {

        @Test
        fun doesNotThrow_whenServerReturnsError() = runTest {
            client.queries.getUserOrErrors("unknown-id")
        }

        @Test
        fun errors_isNonEmpty_forUnknownId() = runTest {
            val res: GetUserOrErrorsFullResponse = client.queries.getUserOrErrors("unknown-id")
            assertNotNull(res.errors)
            assertTrue(res.errors!!.isNotEmpty())
        }

        @Test
        fun errors_hasMessage_forUnknownId() = runTest {
            val res = client.queries.getUserOrErrors("unknown-id")
            val msg = res.errors!![0].message
            assertNotNull(msg)
            assertTrue(msg.isNotBlank())
        }
    }

    // ── findUserOrErrors — null return (no error) ─────────────────────────────

    @Nested
    inner class FindUserOrErrorsNullReturn {

        @Test
        fun doesNotThrow_forUnknownId() = runTest {
            client.queries.findUserOrErrors("user-99")
        }

        @Test
        fun data_fieldIsNull_forUnknownId() = runTest {
            val res: FindUserOrErrorsFullResponse = client.queries.findUserOrErrors("user-99")
            assertTrue(res.data == null || res.data!!.findUserOrErrors == null)
        }

        @Test
        fun errors_isNullOrEmpty_forUnknownId() = runTest {
            val res = client.queries.findUserOrErrors("user-99")
            assertTrue(res.errors == null || res.errors!!.isEmpty())
        }
    }

    // ── findUserOrErrors — server error ───────────────────────────────────────

    @Nested
    inner class FindUserOrErrorsWithErrors {

        @Test
        fun doesNotThrow_forErrorId() = runTest {
            client.queries.findUserOrErrors("error-id")
        }

        @Test
        fun errors_isNonEmpty_forErrorId() = runTest {
            val res: FindUserOrErrorsFullResponse = client.queries.findUserOrErrors("error-id")
            assertNotNull(res.errors)
            assertTrue(res.errors!!.isNotEmpty())
        }
    }

    // ── findUserOrErrors — success ────────────────────────────────────────────

    @Nested
    inner class FindUserOrErrorsSuccess {

        @Test
        fun data_isPresent_forKnownId() = runTest {
            val res: FindUserOrErrorsFullResponse = client.queries.findUserOrErrors("user-1")
            assertNotNull(res.data)
            assertNotNull(res.data!!.findUserOrErrors)
            assertEquals("user-1", res.data!!.findUserOrErrors!!.id)
        }
    }

    // ── Non-captureErrors query throws on server error ────────────────────────

    @Nested
    inner class NonCaptureErrorsThrows {

        @Test
        fun getUser_throws_forUnknownId() = runTest {
            assertFailsWith<Exception> {
                client.queries.getUser("unknown-id")
            }
        }
    }

    // ── createUserOrErrors mutation ───────────────────────────────────────────

    @Nested
    inner class CreateUserOrErrorsMutation {

        @Test
        fun success_dataIsPresent() = runTest {
            val res: CreateUserOrErrorsFullResponse = client.mutations.createUserOrErrors(kCreateInput)
            assertNotNull(res.data)
            assertNotNull(res.data!!.createUserOrErrors)
        }

        @Test
        fun success_noErrors() = runTest {
            val res = client.mutations.createUserOrErrors(kCreateInput)
            assertTrue(res.errors == null || res.errors!!.isEmpty())
        }

        @Test
        fun doesNotThrow_whenServerReturnsError() = runTest {
            client.mutations.createUserOrErrors(kErrorInput)
        }

        @Test
        fun errors_arePopulated_whenServerReturnsError() = runTest {
            val res: CreateUserOrErrorsFullResponse = client.mutations.createUserOrErrors(kErrorInput)
            assertNotNull(res.errors)
            assertTrue(res.errors!!.isNotEmpty())
        }
    }
}
